prov/rxd/src/rxd_av.c

```diff
@@ -251,35 +231,49 @@ static int rxd_av_remove(struct fid_av *av_fid, fi_addr_t *fi_addr, size_t count
 			uint64_t flags)
 {
 	int ret = 0;
-	size_t i, addrlen;
+	size_t i;
 	fi_addr_t rxd_addr;
 	struct rxd_av *av;
-	uint8_t addr[RXD_NAME_LENGTH];
+    struct dlist_entry *av_entry;
+    struct util_ep *util_ep;
+    struct rxd_ep *rxd_ep;
+    struct rxd_peer *peer_entry;
+    struct rxd_fiaddr_entry *entry;
 
 	av = container_of(av_fid, struct rxd_av, util_av.av_fid);
 	fastlock_acquire(&av->util_av.lock);
 	for (i = 0; i < count; i++) {
-		rxd_addr = av->fi_addr_table[fi_addr[i]];
 
-		addrlen = RXD_NAME_LENGTH;
-		ret = fi_av_lookup(av->dg_av, av->rxd_addr_table[rxd_addr].dg_addr,
-				   addr, &addrlen);
-		if (ret)
-			goto err;
-		
-		ret = ofi_rbmap_find_delete(&av->rbmap, (void *) addr);
-		if (ret)
-			goto err;
-
-		ret = fi_av_remove(av->dg_av, &av->rxd_addr_table[rxd_addr].dg_addr,
-				   1, flags);
-		if (ret)
-			goto err;
-
-		av->fi_addr_table[fi_addr[i]] = FI_ADDR_UNSPEC;
-		av->rxd_addr_table[rxd_addr].fi_addr = FI_ADDR_UNSPEC;
-		av->rxd_addr_table[rxd_addr].dg_addr = FI_ADDR_UNSPEC;
-		av->dg_av_used--;
+        // remove mappings between fi_addr --> dg_addr
+        //  remove each endpoint hash mapping between 
+        // fi_addr --> rxd_addr.
+        // reset peer_entry->fi_addr value to UNSPEC
+        // do not delete the peer entry. it may be being used 
+        // for an underlying existing communication. 
+        // donot delete dg_addr from dg_av.
+        ret =  ofi_av_remove_addr(&av->util_av, fi_addr[i]);
+        if (ret)
+            goto err;
+
+        dlist_foreach(&av->util_av.ep_list, av_entry) {
+
+            HASH_FIND(hh, rxd_ep->fi_rxdaddr_hash, (void*)(&fi_addr[i]), 
+                        sizeof(fi_addr_t), entry);
+            if(!entry)
+                goto err;
+            rxd_addr = entry->rxd_addr;
+            util_ep = container_of(av_entry, struct util_ep, av_entry);
+            rxd_ep =  container_of(util_ep, struct rxd_ep, util_ep);
+
+            // function to get peer_entry from ep->rxd_peer_pool
+            peer_entry = ofi_bufpool_get_ibuf(rxd_ep->peer_pool.pool, rxd_addr);
+            assert(peer_entry->peer_addr == rxd_addr);
+    
+            peer_entry->fi_addr = FI_ADDR_UNSPEC;
+            
+            HASH_DELETE(hh, rxd_ep->fi_rxdaddr_hash, entry);
+            free(entry);
```

Comment:
On a second look i feel that the setting and unsetting of fi_addr in the peer entry is redundant I don't see peer_entry->fi_addr being used anywhere in the code. since we always use the fi_addr-->rxd_addr to fetch the peer entry we cannot use the fi_addr stored inside the peer entry.  Also it seems that the removal of the fi_addr-->rxd_addr mapping in the rxd_av_remove as it may cause issues in the underlying protocol for fetching the peer if some communication is still underway?
rxd_av_remove